File: subsync/synchro/pipeline.py
import gizmo
from subsync.synchro import speech
from subsync.data import languages
from subsync.error import Error
import math

import logging
logger = logging.getLogger(__name__)


class BasePipeline(object):
    def __init__(self, stream, runCb=None):
        self.demux = gizmo.Demux(stream.path, runCb)
        self.extractor = gizmo.Extractor(self.demux)

        self.duration = max(self.demux.getDuration(), 0)
        self.timeWindow = [0, self.duration]

    def destroy(self):
        self.extractor.stop()
        self.extractor.wait()
        self.extractor.connectEosCallback(None)
        self.extractor.connectErrorCallback(None)

    def selectTimeWindow(self, begin, end=math.inf):
        if begin != 0.0:
            self.demux.seek(begin)
        if end is not math.inf:
            self.extractor.selectEndTime(end)
        else:
            end = self.duration
        self.timeWindow = (begin, end)

    def start(self, threadName=None):
        self.extractor.start(threadName=threadName)

    def stop(self):
        self.extractor.stop()

    def isRunning(self):
        return self.extractor.isRunning()

    def getProgress(self):
        if self.duration:
            pos = self.demux.getPosition()
            if math.isclose(pos, 0.0):
                pos = self.timeWindow[0]
            if self.timeWindow[1] > self.timeWindow[0]:
                return (pos - self.timeWindow[0]) / (self.timeWindow[1] - self.timeWindow[0])

    def getPosition(self):
        return max(self.timeWindow[0], min(self.demux.getPosition(), self.timeWindow[1]))

    def connectErrorCallback(self, cb):
        self.extractor.connectErrorCallback(cb)

    def connectEosCallback(self, cb):
        self.extractor.connectEosCallback(cb)


class SubtitlePipeline(BasePipeline):
    def __init__(self, stream, runCb=None):
        ''' Speech recognition pipeline:

        Demux --> SpeechDec --[words]--> {NgramSplitter} --[words]--> ...
        '''

        super().__init__(stream, runCb)
        self.dec = gizmo.SubtitleDec()
        self.ngramSplitter = None
        self.sink = self.dec

        langInfo = stream.lang and languages.get(code3=stream.lang.lower())
        if langInfo:
            if langInfo.rightToLeft:
                logger.info('switching to right-to-left for file "%s"', stream.path)
                self.dec.setRightToLeft(True)

            if langInfo.ngrams:
                logger.info('switching to %i-gram for file "%s"', langInfo.ngrams, stream.path)
                self.dec.setMinWordLen(langInfo.ngrams)
                self.ngramSplitter = gizmo.NgramSplitter(langInfo.ngrams)
                self.dec.addWordsListener(self.ngramSplitter.pushWord)
                self.sink = self.ngramSplitter

        if stream.enc != None:
            self.dec.setEncoding(stream.enc)

        self.demux.connectDec(self.dec, stream.no)

    def configure(self, minWordLen, minWordProb=None):
        self.dec.setMinWordLen(minWordLen)

    def destroy(self):
        super().destroy()
        self.dec.removeWordsListener()
        self.dec.removeSubsListener()
        self.sink.removeWordsListener()

    def addWordsListener(self, listener):
        self.sink.addWordsListener(listener)

    def removeWordsListener(self, listener=None):
        return self.sink.removeWordsListener(listener)

    def getRawWordsSource(self):
        return self.dec

    def addSubsListener(self, listener):
        self.dec.addSubsListener(listener)

    def removeSubsListener(self, listener=None):
        self.dec.removeSubsListener(listener)


class SpeechPipeline(BasePipeline):
    def __init__(self, stream, runCb=None):
        ''' Speech recognition pipeline:

        Demux --> AudioDec --> Resampler --> SpeechRecognition --[words]--> {NgramSplitter} --[words]--> ...
        '''

        super().__init__(stream, runCb)

        speechModel = speech.loadSpeechModel(stream.lang)
        self.dec = gizmo.AudioDec()

        speechAudioFormat = speech.getSpeechAudioFormat(speechModel)
        logger.info('speech recognition audio format: %s', speechAudioFormat)

        self.speechRec = speech.createSpeechRec(speechModel)
        self.ngramSplitter = None
        self.sink = self.speechRec

        langInfo = stream.lang and languages.get(code3=stream.lang.lower())
        if langInfo and langInfo.ngrams:
            logger.info('switching to %i-gram for audio "%s"', langInfo.ngrams, stream.path)
            self.speechRec.setMinWordLen(langInfo.ngrams)
            self.ngramSplitter = gizmo.NgramSplitter(langInfo.ngrams)
            self.speechRec.addWordsListener(self.ngramSplitter.pushWord)
            self.sink = self.ngramSplitter

        self.resampler = gizmo.Resampler()
        self.channels = stream.channels
        self.resampler.connectFormatChangeCallback(self.onAudioFormatChanged)

        self.demux.connectDec(self.dec, stream.no)
        self.dec.connectOutput(self.resampler)
        self.resampler.connectOutput(self.speechRec, speechAudioFormat)

    def configure(self, minWordLen, minWordProb):
        self.speechRec.setMinWordProb(minWordProb)
        self.speechRec.setMinWordLen(minWordLen)

    def destroy(self):
        super().destroy()
        self.sink.removeWordsListener()
        self.speechRec.removeWordsListener()
        self.resampler.connectFormatChangeCallback(None)

    def onAudioFormatChanged(self, inFormat, outFormat):
        logger.info('input audio format: %s', inFormat)
        channelsMap = self.channels.getLayoutMap(inFormat.channelLayout)
        logger.info('listening to channels: %s', channelsMap)
        self.resampler.setChannelMap(channelsMap.getMap())

    def addWordsListener(self, listener):
        self.sink.addWordsListener(listener)

    def removeWordsListener(self, listener=None):
        return self.sink.removeWordsListener(listener)

    def getRawWordsSource(self):
        return self.speechRec


def createProducerPipeline(stream, runCb=None):
    if stream.type == 'subtitle/text':
        return SubtitlePipeline(stream, runCb)
    elif stream.type == 'audio':
        return SpeechPipeline(stream, runCb)
    else:
        raise Error(_('Not supported stream type'), type=stream.type)


def createProducerPipelines(stream, no=None, runCb=None):
    pipes = []
    for i in range(no):
        p = createProducerPipeline(stream, runCb)
        pipes.append(p)

        if p.duration:
            if p.duration / (i+1) < 5*60:
                logger.info('using only %i jobs due to short duration', i+1)
                break
        else:
            logger.warn('cannot get duration - using single pipeline')
            break

    if len(pipes) > 1:
        no = len(pipes)
        duration = pipes[0].duration
        partTime = duration / no
        for i, p in enumerate(pipes):
            begin = i * partTime
            end = begin + partTime + 1.0 # overlap to catch whole words
            if end >= duration:
                end = math.inf
            logger.info('job %i/%i time window set to %.2f - %.2f', i+1, no, begin, end)

            try:
                p.selectTimeWindow(begin, end)
            except gizmo.Error:
                logger.warning('seek failed, using only %i jobs', i, exc_info=True)
                no = i
                pipes = pipes[:no]
                if pipes:
                    logger.info('updating job %i/%i time window set to %.2f - %.2f',
                            no, no, begin - partTime, duration)
                    pipes[-1].selectTimeWindow(begin - partTime, duration)
                break

    return pipes

File: web/src/extractor/pipeline.js
import Gizmo from '../gizmowrap.js';
import Assets from './assets.js';
import settings from '../settings.js';
import languages from '../data/languages.json';
import Logger from '../logger.js';
const logger = Logger.logger.get('[Pipeline]');

export default class Pipeline {

  constructor(path) {
    this.demux = new Gizmo.instance.Demux(path);
    this.streams = this.demux.getStreamsInfo();
    const lang = getLangFromPath(path);
    if (lang) {
      for (const stream of this.streams) {
        stream.lang = stream.lang || lang;
      }
    }
  }

  makePipeline(stream, params, path) {
    let output = null;
    if (stream.type === 'audio') {
      output = this.makeSpeechPipeline(stream);
    } else if (stream.type === 'subtitle/text') {
      output = this.makeSubPipeline(stream, path);
    }

    const langInfo = stream.lang && languages.find(lang => lang.code3 === stream.lang);
    if (langInfo && langInfo.ngrams) {
      logger.log(`switching to ${langInfo.ngrams}-gram for language ${langInfo.name}`);
      output.setMinWordLen(langInfo.ngrams);
      this.ngramSplitter = new Gizmo.instance.NgramSplitter(langInfo.ngrams);
      output.connectNgramSplitter(this.ngramSplitter);
      output = this.ngramSplitter;
    }

    if (stream.lang && params.otherLang && stream.lang !== params.otherLang) {
      try {
        var dict = Assets.loadDictionary(stream.lang, params.otherLang, settings.minWordLen);
        logger.log(`loaded dict with ${dict.size()} entries`);
        this.translator = new Gizmo.instance.Translator(dict);
        this.translator.setMinWordsSim(settings.minWordsSim);
        output.connectTranslator(this.translator);
        output = this.translator;
      } finally {
        dict && dict.delete();
      }
    }

    return output;
  }

  makeSubPipeline(stream, path) {
    this.dec = new Gizmo.instance.SubtitleDec();
    this.dec.setMinWordLen(settings.minWordLen);
    const enc = stream.enc || this.detectCharEncoding(stream, path);
    if (enc) {
      this.dec.setEncoding(enc);
    }
    const lang = stream.lang && languages.find(lang => lang.code3 === stream.lang);
    if (lang && lang.rightToLeft) {
      logger.log('switching to right-to-left for language', lang.name)
      this.dec.setRightToLeft(true);
    }
    this.demux.connectDec(this.dec, stream.no);
    return this.dec;
  }

  detectCharEncoding(stream, path) {
    if (this.streams.length == 1 && this.streams[0].type === 'subtitle/text') {
      const enc = Gizmo.instance.detectCharEncoding(path, 32*1024*1024);
      logger.log(`detecting character encoding: ${enc}`);
      if (enc === 'ascii') {
        const lang = stream.lang && languages.find(lang => lang.code3 === stream.lang);
        if (lang && lang.enc && lang.enc.length) {
          logger.log(`detected encoding ${lang.enc[0]} for language ${lang.code3}`);
          return lang.enc[0];
        }
      } else if (enc !== 'binary') {
        return enc;
      }
    }
  }

  makeSpeechPipeline(stream) {
    if (!stream.lang) {
      throw {
        message: 'Language not selected',
        file: stream.file && stream.file.name,
      }
    }
    const gizmo = Gizmo.instance;
    const speechModel = Assets.loadSpeechModel(stream.lang);
    logger.log('speech model', speechModel);
    if (!speechModel) {
      throw {
        message: 'Missing speech recognition model',
        file: stream.file && stream.file.name,
        language: stream.lang,
      }
    }

    this.speechRec = new gizmo.SpeechRecognition();
    for (const [key, val] of Object.entries(speechModel.sphinx)) {
      this.speechRec.setParam(key, val);
    }
    this.speechRec.setParam('-mmap', '0');
    this.speechRec.setMinWordProb(settings.minWordProb);
    this.speechRec.setMinWordLen(settings.minWordLen);

    this.dec = new gizmo.AudioDec();
    this.resampler = new gizmo.Resampler();
    this.demux.connectDec(this.dec, stream.no);
    this.dec.connectOutput(this.resampler);

    this.resampler.connectOutput(this.speechRec,
      gizmo.AVSampleFormat[speechModel.sampleformat],
      parseInt(speechModel.samplerate), 32*1024);

    return this.speechRec;
  }

  addSubsListener(listener) {
    this.dec.addSubsListener(listener);
  }

  destroy() {
    if (this.demux) this.demux.delete();
    if (this.dec) this.dec.delete();
    if (this.speechRec) this.speechRec.delete();
    if (this.resampler) this.resampler.delete();
    if (this.ngramSplitter) this.ngramSplitter.delete();
    if (this.translator) this.translator.delete();
    this.demux = undefined;
    this.dec = undefined;
    this.speechRec = undefined;
    this.resampler = undefined;
    this.ngramSplitter = undefined;
    this.translator = undefined;
  }
}

function getLangFromPath(path) {
  const ents = path.split('.');
  const code = ents.length >= 2 && ents[ents.length - 2].match(/[a-zA-Z]{2,3}$/g);
  if (code && code.length) {
    const c = code[0].toLowerCase();
    const lang = languages.find(l => c === l.code3 || c === l.code2 || (l.extraCodes && l.extraCodes.includes(c)));
    return lang && lang.code3;
  }
}
